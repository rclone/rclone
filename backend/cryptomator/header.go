package cryptomator

import (
	"bytes"
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"io"
	"unsafe"
)

// fileHeader is the header of an encrypted Cryptomator file.
type fileHeader struct {
	// The nonce used to encrypt the file header. Each file content chunk, while encrypted with its own nonce, also mixes the file header nonce into the MAC.
	Nonce    []byte
	Reserved []byte
	// The AES key used to encrypt the file contents.
	ContentKey []byte
}

const (
	// headerContentKeySize is the size of the ContentKey in the FileHeader.
	headerContentKeySize = 32
	// headerReservedSize is the size of the Reserved data in the FileHeader.
	headerReservedSize = 8
	// headerPayloadSize is the size of the encrypted part of the file header.
	headerPayloadSize = headerContentKeySize + headerReservedSize
	// headerReservedValue is the expected value of the Reserved data.
	headerReservedValue uint64 = 0xFFFFFFFFFFFFFFFF
)

// NewHeader creates a new randomly initialized FileHeader
func (c *cryptor) NewHeader() (header fileHeader, err error) {
	header.Nonce = make([]byte, c.nonceSize())
	header.ContentKey = make([]byte, headerContentKeySize)
	header.Reserved = make([]byte, headerReservedSize)

	if _, err = rand.Read(header.Nonce); err != nil {
		return
	}

	if _, err = rand.Read(header.ContentKey); err != nil {
		return
	}

	binary.BigEndian.PutUint64(header.Reserved, headerReservedValue)

	return
}

type headerPayload struct {
	Reserved   [headerReservedSize]byte
	ContentKey [headerContentKeySize]byte
}

var _ [0]struct{} = [unsafe.Sizeof(headerPayload{}) - headerPayloadSize]struct{}{}

func copySameLength(dst, src []byte, name string) error {
	if len(dst) != len(src) {
		return fmt.Errorf("incorrect length of %s: expected %d got %d", name, len(dst), len(src))
	}
	copy(dst, src)
	return nil
}

// marshalHeader encrypts the header and writes it in encrypted form to the writer.
func (c *cryptor) marshalHeader(w io.Writer, h fileHeader) (err error) {
	var payload headerPayload
	if err = copySameLength(payload.Reserved[:], h.Reserved, "Reserved"); err != nil {
		return
	}
	if err = copySameLength(payload.ContentKey[:], h.ContentKey, "ContentKey"); err != nil {
		return
	}

	var encBuffer bytes.Buffer
	if err = binary.Write(&encBuffer, binary.BigEndian, &payload); err != nil {
		return
	}
	encPayload := c.encryptChunk(encBuffer.Bytes(), h.Nonce, nil)
	_, err = w.Write(encPayload)
	return
}

// unmarshalHeader reads an encrypted header from the reader and decrypts it.
func (c *cryptor) unmarshalHeader(r io.Reader) (header fileHeader, err error) {
	encHeader := make([]byte, c.nonceSize()+headerPayloadSize+c.tagSize())
	_, err = io.ReadFull(r, encHeader)
	if err != nil {
		return
	}
	nonce := encHeader[:c.nonceSize()]
	encHeader, err = c.decryptChunk(encHeader, nil)
	if err != nil {
		return
	}

	var payload headerPayload
	if err = binary.Read(bytes.NewReader(encHeader), binary.BigEndian, &payload); err != nil {
		return
	}
	header.Nonce = nonce
	header.ContentKey = payload.ContentKey[:]
	header.Reserved = payload.Reserved[:]
	return
}
