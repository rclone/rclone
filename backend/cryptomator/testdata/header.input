{
    "siv_ctrmac": {
        "CipherCombo": "SIV_CTRMAC",
        "Header": "mDCOS8MT0etiyDKaXv8X0cD1ayTewk/Uir95AZO5t4bwy8sQm5yl23Y0XxyQtg43/IQJLyHLJSuVfKLEQuPmncqbbndSiMzU2iS00Mo9lM3/1XCmLJWmig==",
        "EncKey": "ABygFz/iZU0HXYlE5NtYFPgEmKqyWEh4wGjHo0fJgNY=",
        "MacKey": "lwdzbaZqIfibYj8LSjcUAqpOJfdku/qN13f2OZWpxDk="
    },
    "siv_gcm": {
        "CipherCombo": "SIV_GCM",
        "Header": "WSGqVbG3C7X981CdJIiOFFht5ldI6GzWWFhzPSNaKbw8DZcvjoMVB3tiNNRmFTS2K1Y8rBYx0c18Bbj7Tr/gYuseGMY=",
        "EncKey": "gk3ygEvJMA/z4e4QzItHXtQmM2GEHaXN9fm9esQxTcQ=",
        "MacKey": "zUq4ZFL65Fz6w9htPAggqfllEv3OUQDAbZZJc2eIyJ8="
    }
}
