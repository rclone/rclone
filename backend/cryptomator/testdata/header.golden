{
    "siv_ctrmac": {
        "Nonce": "mDCOS8MT0etiyDKaXv8X0Q==",
        "Reserved": "//////////8=",
        "ContentKey": "RgbuFMgbZP051gzgXp8WbQNJ5y50eNyyBL/6qdQNn6c="
    },
    "siv_gcm": {
        "Nonce": "WSGqVbG3C7X981Cd",
        "Reserved": "//////////8=",
        "ContentKey": "dBS1BH/nsGxL5AF6Z0I+P6V5PSelEFarrIylAIqWmlc="
    }
}
